File: src/main/java/jp/gr/java_conf/mizu/yapp/tools/CommandLineParser.java
package jp.gr.java_conf.mizu.yapp.tools;

import java.util.List;
import java.util.Map;
import java.util.Set;

import jp.gr.java_conf.mizu.yapp.util.CollectionUtil;

public class CommandLineParser {
  public final Map<String, Option> options = CollectionUtil.map();
  public final Map<String, Object> optionValues = CollectionUtil.map();  
  public final List<String> values = CollectionUtil.list();
  public final List<String> descriptions = CollectionUtil.list();
  
  public static enum OptionType {INT, STR, NONE}
  public static class Option {
    public final String key;
    public final String name;
    public final OptionType type;
    public final String description;
    
    public Option(String key, String name, OptionType type, String description) {
      this.key = key;
      this.name = name;
      this.type = type;
      this.description = description;
    }    
  }
  public CommandLineParser opt(String key, String name, OptionType type, String description) {
    options.put(name, new Option(key, name, type, description));
    descriptions.add(description);
    return this;
  }
  public void parse(String[] commandLine) {
    for(int i = 0; i < commandLine.length; i++) {
      Option option = options.get(commandLine[i]);
      if(option == null) {
        values.add(commandLine[i]);
        continue;
      }
      if(option.type == OptionType.NONE) {
        //put dummy object
        optionValues.put(option.key, new Object());
        continue;
      }
      if(i >= commandLine.length - 1) {
        throw new CommandLineException("no argument is supplied for " + option.name);
      }
      i++;
      switch(option.type) {
      case INT:
        try {
          optionValues.put(option.key, Integer.parseInt(commandLine[i]));
        }catch(NumberFormatException e) {
          throw new CommandLineException("argument for " + option.name + " must be integer");
        }
        break;
      case STR:
        optionValues.put(option.key, commandLine[i]);
        break;
      default:
        throw new AssertionError("should not reach here");
      }
    }
  }
  
  public boolean hasOption(String key) {
    return optionValues.containsKey(key);
  }
  
  public String getString(String key) {
    return (String)optionValues.get(key);
  }
  
  public String getString(String key, String defaultValue) {
    return optionValues.containsKey(key) ?
      getString(key) : defaultValue;
  }
  
  public int getInt(String key) {
    return (Integer)optionValues.get(key);
  }
  
  public int getInt(String key, int defaultValue) {
    return optionValues.containsKey(key) ?
      getInt(key) : defaultValue;
  }
}

File: src/main/scala/com/github/kmizu/yapp/tools/CommandLineParser.scala
package com.github.kmizu.yapp.tools

import java.util.List
import java.util.Map
import com.github.kmizu.yapp.util.CollectionUtil

object CommandLineParser {

  sealed abstract class OptionType
  object OptionType {
    case object INT extends OptionType
    case object STR extends OptionType
    case object NONE extends OptionType
  }

  case class Option(key: String, name: String, `type`: CommandLineParser.OptionType, description: String)
}

class CommandLineParser {
  import CommandLineParser.OptionType
  final val options: Map[String, CommandLineParser.Option] = CollectionUtil.map()
  final val optionValues: Map[String, AnyRef] = CollectionUtil.map()
  final val values: List[String] = CollectionUtil.list()
  final val descriptions: List[String] = CollectionUtil.list()

  def opt(key: String, name: String, `type`: CommandLineParser.OptionType, description: String): CommandLineParser = {
    options.put(name, new CommandLineParser.Option(key, name, `type`, description))
    descriptions.add(description)
    this
  }

  def parse(commandLine: Array[String]) {
    {
      var i: Int = 0
      while (i < commandLine.length) {
        val option: CommandLineParser.Option = options.get(commandLine(i))
        if (i >= commandLine.length - 1)  throw new CommandLineException("no argument is supplied for " + option.name)

        if (option == null) {
          values.add(commandLine(i))
        }else if (option.`type` eq OptionType.NONE) {
          optionValues.put(option.key, new AnyRef)
        } else {
          i += 1
          option.`type` match {
            case OptionType.INT =>
              try {
                optionValues.put(option.key, new java.lang.Integer(Integer.parseInt(commandLine(i))))
              } catch {
                case e: NumberFormatException =>
                  throw new CommandLineException("argument for " + option.name + " must be integer")
              }
            case OptionType.STR =>
              optionValues.put(option.key, commandLine(i))
            case _ =>
              throw new AssertionError("should not reach here")
          }
        }
        i += 1
      }
    }
  }

  def hasOption(key: String): Boolean =  optionValues.containsKey(key)

  def getString(key: String): String = optionValues.get(key).asInstanceOf[String]

  def getString(key: String, defaultValue: String): String = {
    if (optionValues.containsKey(key)) getString(key) else defaultValue
  }

  def getInt(key: String): Int = optionValues.get(key).asInstanceOf[java.lang.Integer].intValue()

  def getInt(key: String, defaultValue: Int): Int = if (optionValues.containsKey(key)) getInt(key) else defaultValue
}
